crates/core/src/dht/actions/remove_queued_holding_workflow.rs

```diff
@@ -36,11 +52,36 @@ impl Future for RemoveQueuedHoldingWorkflowFuture {
             .register_waker(self.id.clone(), cx.waker().clone());
 
         if let Some(state) = self.context.try_state() {
-            if state.dht().has_exact_queued_holding_workflow(&self.pending) {
-                Poll::Pending
-            } else {
-                self.context.unregister_waker(self.id.clone());
-                Poll::Ready(())
+            let store = state.dht();
+            match self.state {
+                HoldingWorkflowQueueing::Processing => {
+                    if store.has_exact_queued_holding_workflow(&self.pending)
+                        || !store.has_exact_in_process_holding_workflow(&self.pending)
+                    {
+                        Poll::Pending
+                    } else {
+                        self.context.unregister_waker(self.id.clone());
+                        Poll::Ready(())
+                    }
+                }
+                HoldingWorkflowQueueing::Waiting(_) => {
+                    if !store.has_exact_queued_holding_workflow(&self.pending)
+                        || store.has_exact_in_process_holding_workflow(&self.pending)
+                    {
+                        Poll::Pending
+                    } else {
+                        self.context.unregister_waker(self.id.clone());
+                        Poll::Ready(())
+                    }
+                }
```

Comment:
Looks like these branches are identical?

```suggestion
                HoldingWorkflowQueueing::Processing | HoldingWorkflowQueueing::Waiting(_) => {
                    if store.has_exact_queued_holding_workflow(&self.pending)
                        || !store.has_exact_in_process_holding_workflow(&self.pending)
                    {
                        Poll::Pending
                    } else {
                        self.context.unregister_waker(self.id.clone());
                        Poll::Ready(())
                    }
                }
```

Reply:
Nope, just not well refactored, the ! are on opposite queues because the operation is done when the item moves from one queue to the other.

Reply:
Whoops :)